driver/src/driver/scheduler.rs

```diff
@@ -0,0 +1,274 @@
+    }
+
+    pub fn currently_being_solved(now: SystemTime) -> std::result::Result<Self, SystemTimeError> {
+        Self::current(now).map(|batch_id| batch_id.prev())
+    }
+
+    pub fn order_collection_start_time(self) -> SystemTime {
+        SystemTime::UNIX_EPOCH + Duration::from_secs(self.0 * BATCH_DURATION.as_secs())
+    }
+
+    pub fn solve_start_time(self) -> SystemTime {
+        self.order_collection_start_time() + BATCH_DURATION
+    }
+
+    pub fn next(self) -> BatchId {
+        self.0.checked_add(1).map(BatchId).unwrap()
+    }
+
+    pub fn prev(self) -> BatchId {
+        self.0.checked_sub(1).map(BatchId).unwrap()
+    }
+}
+
+#[derive(Debug)]
+pub struct AuctionTimingConfiguration {
+    /// The offset from the start of a batch at which point we should start
+    /// solving.
+    pub target_start_solve_time: Duration,
+    /// The offset from the start of the batch at which point there is not
+    /// enough time left to attempt to solve.
+    pub latest_solve_attempt_time: Duration,
+}
+
+pub struct Scheduler<'a> {
+    driver: &'a mut dyn StableXDriver,
+    auction_timing_configuration: AuctionTimingConfiguration,
+    last_solved_batch: Option<BatchId>,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+enum Action {
+    Solve(BatchId),
+    Sleep(Duration),
+}
+
+impl<'a> Scheduler<'a> {
+    pub fn new(
+        driver: &'a mut dyn StableXDriver,
+        auction_timing_configuration: AuctionTimingConfiguration,
+    ) -> Self {
+        assert!(auction_timing_configuration.latest_solve_attempt_time <= BATCH_DURATION);
+        assert!(
+            auction_timing_configuration.target_start_solve_time
+                < auction_timing_configuration.latest_solve_attempt_time
+        );
+        Self {
+            driver,
+            auction_timing_configuration,
+            last_solved_batch: None,
+        }
+    }
+
+    pub fn run_forever(&mut self) {
+        loop {
+            let now = SystemTime::now();
+            match self.determine_action(now) {
+                Ok(Action::Sleep(duration)) => {
+                    info!("Sleeping {}s.", duration.as_secs());
+                    thread::sleep(duration);
+                }
+                Ok(Action::Solve(batch_id)) => {
+                    info!("Solving batch {}.", batch_id.0);
+                    self.run_solver(batch_id);
+                }
+                Err(err) => {
+                    error!("Scheduler error: {}", err);
+                    thread::sleep(Duration::from_secs(10));
+                }
+            }
+        }
+    }
+
+    fn run_solver(&mut self, batch_id: BatchId) {
+        info!("Running solver for batch {}.", batch_id.0);
+        if let Err(err) = self.driver.run(batch_id.0.into()) {
+            error!("StableXDriver error: {}", err);
+            thread::sleep(Duration::from_secs(10));
+        } else {
+            self.last_solved_batch.replace(batch_id);
```

Comment:
In the current logic we also skip the batch going forward if the error came from the price finding, verification or solution submission as in those case it's unlikely to succeed in a retry (and it might clog the next batch). So really we only retry if the order book fetching failed for some reason (that was handy when we were running into occasional timeouts).

In order to reproduce this behaviour we would have to differentiate between driver error on a more fine level (maybe a should_retry flag or the like) and I'm not sure that's worth it. So I'd suggest to always add the batchID as "last_solved_batch" even if there was an error.

What do you think?

Reply:
I agree as well, we should log the error, and skip to the next batch in case of solver error. The `driver.run()` method can return something like a:
```
enum DriverResult {
    Ok,
    Retry(Error),
    Skip(Error),
}
```

That way we can differentiate between an error where we want to retry and a hard error where we want to skip the batch.

Reply:
Implemented this.